packages/flutter_tools/lib/src/commands/daemon.dart

```diff
@@ -456,9 +457,15 @@ class AppDomain extends Domain {
     if (app == null)
       throw "app '$appId' not found";
 
-    return app._runInZone(this, () {
+    if (!fullRestart && inProgressHotReload != null)
+      throw 'hot restart already in progress';
+
+    final Future<OperationResult> action = app._runInZone(this, () {
       return app.restart(fullRestart: fullRestart, pauseAfterRestart: pauseAfterRestart);
     });
+    if (!fullRestart)
+      inProgressHotReload = action.then((_) => inProgressHotReload = null);
+    return action;
```

Comment:
It may be possible to simplify this all to:

```
_inProgressReload = app._runInZone(this, () {
  return app.restart(fullRestart: fullRestart, pauseAfterRestart: pauseAfterRestart);
});

return _inProgressReload.whenComplete(() {
  _inProgressReload = nu'l;
});
```

In which case we'd want to type `_inProgressReload` as `Future<OperationResult>`. I think the above will have the nice benefit that errors in the reload operation future will be passed back to the caller (and back over the wire).

Reply:
Oops, good catch - I think my original code would've failed to set the in-progress future back to null if the `runInZone` future errorred. 2a58cd8 changes to your code (similar to the earlier change, this removes the checking for only hot reloads, so the future will be set for full restarts too, but I think you intended this).